[Merged by Bors] - feat(algebra/module/graded_module): define graded module

jjaassoonn opened the pull request:
By imitating the current `graded_algebra`, this pr builds `graded_module` over some `graded algebra`

Co-authored-by: Eric Wieser @eric-wieser 

- [x] depends on: #14626
- [x] depends on: #15654

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

jjaassoonn committed:
initial

jjaassoonn committed:
add doc

jjaassoonn committed:
remove lint

eric-wieser commented on src/algebra/module/graded_module.lean:
```diff
@@ -0,0 +1,136 @@
+# Graded Module
+
+Given a graded `R`-algebra `A` with grading `𝓐 : ι → submodule R A` and an `A`-module `M` with
+decomposition `ι → addsubmonoid M`,
```
We need to be a bit careful here; some references define a graded module as what you'd call a graded module over a graded ring, with `R i * M j <= M (i + j)`.

eric-wieser replied:
Oh nevermind, that's exactly what you've written here! In that case, my objection is the opposite; I think Bourbaki uses the simpler definition.

jjaassoonn replied:
Is that `smul_mem` below

eric-wieser commented on src/algebra/module/graded_module.lean:
```diff
@@ -0,0 +1,136 @@
+(decompose' : M → ⨁ i, 𝓜 i)
+(left_inv : function.left_inverse decompose' (direct_sum.coe_add_monoid_hom 𝓜))
+(right_inv : function.right_inverse decompose' (direct_sum.coe_add_monoid_hom 𝓜))
+(smul_mem : ∀ ⦃i j : ι⦄ {a : A} {m : M} (hi : a ∈ 𝓐 i) (hj : m ∈ 𝓜 j), a • m ∈ 𝓜 (i + j))
```
This field looks to me like it should exist as a standalone `set_like.graded_smul` class to match `set_like.graded_monoid`

jjaassoonn replied:
I have added type class for `set_like.has_graded_smul` @eric-wieser

eric-wieser commented on src/algebra/module/graded_module.lean:
```diff
@@ -0,0 +1,136 @@
+class graded_module :=
+(decompose' : M → ⨁ i, 𝓜 i)
+(left_inv : function.left_inverse decompose' (direct_sum.coe_add_monoid_hom 𝓜))
+(right_inv : function.right_inverse decompose' (direct_sum.coe_add_monoid_hom 𝓜))
```
I'm beginning to think we want to extract this to a `direct_sum.has_decomposition` typeclass, rather than repeating it here and in `graded_algebra`.

jjaassoonn replied:
This sounds like a good idea, may be we should start another pr doing this before #14583

eric-wieser replied:
I think they can probably happen in either order.

eric-wieser replied:
Done in #14626; I think you were right that it makes sense to do that step first.

jjaassoonn replied:
Let me change this pr accordingly.

eric-wieser committed:
feat(algebra/direct_sum/decomposition): add decompositions into a direct sum

This is a constructive version of `direct_sum.is_internal`, and generalizes the existing `graded_algebra`.

eric-wieser committed:
fix

eric-wieser committed:
lintfix

eric-wieser committed:
fix test

jjaassoonn committed:
Merge remote-tracking branch 'origin/eric-wieser/direct_sum.decomposition' into jjaassoonn/graded_module

jjaassoonn committed:
changed to new spelling

jjaassoonn committed:
fix

eric-wieser committed:
remove graded_algebra.support

eric-wieser committed:
Fix lemma names

eric-wieser committed:
Fix

eric-wieser committed:
Fix build errors

jjaassoonn committed:
fix

jjaassoonn committed:
Merge remote-tracking branch 'origin/eric-wieser/direct_sum.decomposition' into jjaassoonn/graded_module

eric-wieser committed:
golf

jjaassoonn committed:
Merge remote-tracking branch 'origin/eric-wieser/direct_sum.decomposition' into jjaassoonn/graded_module

jjaassoonn committed:
Merge remote-tracking branch 'origin/master' into jjaassoonn/graded_module

jjaassoonn committed:
add linear_equiv

jjaassoonn committed:
add set_like.has_graded_smul

eric-wieser commented on src/algebra/module/graded_module.lean:
```diff
@@ -0,0 +1,132 @@
+`M` into `𝓜 : ι → add_submonoid M`, we say that `M` is graded by `𝓜` if and only if the
+decomposition map `M → ⨁ i, 𝓜 i` is inverse to the canonical map `⨁ i, 𝓜 i → M`.
+-/
+class graded_module extends direct_sum.decomposition 𝓜, set_like.has_graded_smul 𝓐 𝓜
```
This warning is real; `graded_module` can't extend `direct_sum.decomposition` for the same reason that `module` can't extend `add_monoid`; the decomposition of `𝓜` doesn't even mention `𝓐`, so this instance would never be found.

The solution is to just use `[direct_sum.decomposition 𝓜] [set_like.has_graded_smul 𝓐 𝓜]` directly.

jjaassoonn replied:
Thanks, I was really confused about this warning message. Now is fixed.

eric-wieser commented on src/algebra/module/graded_module.lean:
```diff
@@ -0,0 +1,132 @@
+@[simp] lemma decompose_symm_of [graded_module 𝓐 𝓜] {i : ι} (x : 𝓜 i) :
+  (graded_module.decompose 𝓐 𝓜).symm (direct_sum.of _ i x) = x :=
+direct_sum.coe_add_monoid_hom_of _ _ _
+
```
I think you should just remove all these, since they just add new ways to spell things we already have.

jjaassoonn replied:
done

eric-wieser commented on src/algebra/module/graded_module.lean:
```diff
@@ -0,0 +1,132 @@
+  (graded_module.decompose 𝓐 𝓜).symm (direct_sum.of _ i x) = x :=
+direct_sum.coe_add_monoid_hom_of _ _ _
+
+instance self : graded_module 𝓐 (λ i, (𝓐 i).to_add_submonoid) :=
```
Maybe `graded_algebra.to_graded_module` would be a good name here to match `semiring.to_module`?

jjaassoonn replied:
done

eric-wieser commented on src/algebra/module/graded_module.lean:
```diff
@@ -0,0 +1,132 @@
+variables [decidable_eq ι] [add_monoid ι] [comm_semiring R] [semiring A] [algebra R A]
+variables (𝓐 : ι → submodule R A) [graded_algebra 𝓐]
+variables {M : Type*} [add_comm_monoid M] [module A M]
+variables (𝓜 : ι → add_submonoid M)
```
If possible, this should use `add_submonoid_class` like `decomposition` does.

jjaassoonn replied:
done

jjaassoonn committed:
fix

jjaassoonn committed:
delete outdated doc

eric-wieser commented on src/algebra/module/graded_module.lean:
```diff
@@ -0,0 +1,103 @@
+`⨁ i, 𝓜 i` is also an `A`-module, via `a • z = decompose (a • redecompose z)` where `decompose` and
+`recompose` are the cannonical homomorphism `M → ⨁ i, 𝓜 i` and `⨁ i, 𝓜 i → M`.
+-/
+def is_module [graded_algebra 𝓐] [@direct_sum.decomposition ι M σ _ _ _ _ 𝓜] :
```
Is the `@` really needed here?

jjaassoonn replied:
Yes, Lean struggles to find `add_submonoid_class σ ?m_1` and `set_like σ ?m_1`.

eric-wieser commented on src/algebra/module/graded_module.lean:
```diff
@@ -0,0 +1,103 @@
+-/
+def is_module [graded_algebra 𝓐] [@direct_sum.decomposition ι M σ _ _ _ _ 𝓜] :
+  module A (⨁ i, 𝓜 i) :=
+{ smul := λ a z, direct_sum.decompose_add_equiv 𝓜 (a • (direct_sum.decompose_add_equiv 𝓜).symm z),
```
I think you can do this without needing a decomposition on M, by summing over the decomposition on A.

I would be tempted to copy the `graded_monoid.ghas_mul` class for `smul` and see where that leads. Mostly that would give you graded multiplicative actions too.

jjaassoonn replied:
I have added `ghas_scalar` classes

jjaassoonn replied:
I have changed definition of scalar multiplication to essentially this:
```lean
def scalar_hom [decidable_eq ι]
   [@set_like.has_graded_scalar ι _ A _ M _ _ _ _ 𝓐 𝓜] :
   (⨁ i, 𝓐 i) →+ (⨁ i, 𝓜 i) →+ ⨁ i, 𝓜 i :=
 direct_sum.to_add_monoid $ λ i,
   add_monoid_hom.flip $ direct_sum.to_add_monoid $ λ j, add_monoid_hom.flip $
     (direct_sum.of (λ i, 𝓜 i) _).comp_hom.comp $ gscalar_hom 𝓐 𝓜
```
Is this what you have in mind @eric-wieser

eric-wieser replied:
Yes, that's precisely what I had in mind

jjaassoonn committed:
add graded_monoid.ghas_scalar

jjaassoonn committed:
change

jjaassoonn committed:
change

eric-wieser commented on src/algebra/direct_sum/basic.lean:
```diff
@@ -76,6 +76,14 @@ def mk (s : finset ι) : (Π i : (↑s : set ι), β i.1) →+ ⨁ i, β i :=
+begin
+  induction h1,
+  exact eq.rec_on (eq_of_heq h2) rfl,
+end
```
Does `by congr'` solve this?

jjaassoonn replied:
No, both `congr` and `congr'` doesn't solve this

jjaassoonn replied:
`by substs h1 h2` this does!

jjaassoonn committed:
golf

jjaassoonn committed:
update docs

jjaassoonn committed:
add doc

jjaassoonn committed:
minor golf

jjaassoonn committed:
less @symbol

jjaassoonn committed:
less @symbol

eric-wieser commented on src/algebra/module/graded_module.lean:
```diff
@@ -0,0 +1,249 @@
+
+variables {ι R A M σ : Type*}
+variables [add_monoid ι] [comm_semiring R] [semiring A] [algebra R A]
+variables (𝓐 : ι → submodule R A)
```
Now that `graded_ring` is merged, you should be able to change this to
```suggestion
variables (𝓐 : ι → σ') [set_like σ' A] [add_submonoid_class σ' A]
```
or similar

jjaassoonn replied:
done

eric-wieser commented on src/algebra/direct_sum/basic.lean:
```diff
@@ -76,6 +76,11 @@ def mk (s : finset ι) : (Π i : (↑s : set ι), β i.1) →+ ⨁ i, β i :=
 def of (i : ι) : β i →+ ⨁ i, β i :=
 dfinsupp.single_add_hom β i
 
+lemma of_congr {i i' : ι} (h1 : i = i') {x : β i} {y : β i'}
```
Elsewhere I've been using [`direct_sum.of_eq_of_graded_monoid_eq`](https://leanprover-community.github.io/mathlib_docs/algebra/direct_sum/ring.html#direct_sum.of_eq_of_graded_monoid_eq) instead of needing this lemma.

jjaassoonn replied:
done

jjaassoonn committed:
Merge remote-tracking branch 'origin/master' into jjaassoonn/graded_module

jjaassoonn committed:
generalize

jjaassoonn committed:
change according to Eric

eric-wieser commented on src/algebra/module/graded_module.lean:
```diff
@@ -0,0 +1,257 @@
+lemma graded_algebra.to_graded_module [decidable_eq ι]
+  [graded_ring 𝓐] :
+  set_like.has_graded_scalar 𝓐 (λ i, (𝓐 i)) :=
+{ smul_mem := λ i j x y hi hj, set_like.graded_monoid.mul_mem hi hj }
```
This looks like it should be called `set_like.graded_monoid.to_has_graded_scalar` to me, and belongs in the other file.

eric-wieser commented on src/algebra/graded_monoid.lean:
```diff
@@ -113,11 +113,20 @@ instance ghas_one.to_has_one [has_zero ι] [ghas_one A] : has_one (graded_monoid
 
+/-- A graded version of `has_scalar`. Scalar multiplication combines grades additively, i.e.
+if `a ∈ A i` and `m ∈ M j`, then `a • b` must be in `M (i + j)`-/
+class ghas_scalar [has_add ι] :=
```
I think we should probably have a `gmodule` class too; and if you're feeling generous, the intermediate `gmul_action` and `gdistrib_mul_action`.

eric-wieser replied:
(we don't need that for `set_like`, only for the dependent versions)

jjaassoonn replied:
I am not actually sure if I am doing what you are asking for, please check the `gmodule` class to see if it is what you want it to do

jjaassoonn committed:
Merge remote-tracking branch 'origin/master' into jjaassoonn/graded_module

jjaassoonn committed:
apply suggestion from code review @eric-wieser

eric-wieser commented on src/algebra/module/graded_module.lean:
```diff
@@ -0,0 +1,353 @@
+    ext,
+    change (0 : A) • (b : M) = 0,
+    rw zero_smul,
+  end }
```
```suggestion
instance gmodule [decidable_eq ι] : gmodule (λ i, 𝓐 i) (λ i, 𝓜 i) :=
{ smul := λ i j x y, ⟨(x : A) • (y : M), set_like.has_graded_smul.smul_mem x.2 y.2⟩,
  one_smul := λ ⟨i, m⟩, sigma.subtype_ext (zero_add _) (one_smul _ _),
  mul_smul := λ ⟨i, a⟩ ⟨j, a'⟩ ⟨k, b⟩, sigma.subtype_ext (add_assoc _ _ _) (mul_smul _ _ _),
  smul_add := λ i j a b c, subtype.ext $ smul_add _ _ _,
  smul_zero := λ i j a, subtype.ext $ smul_zero _,
  add_smul := λ i j a a' b, subtype.ext $ add_smul _ _ _,
  zero_smul := λ i j b, subtype.ext $ zero_smul _ _ }
```

jjaassoonn committed:
golf

jjaassoonn committed:
fix

jjaassoonn committed:
fix

jjaassoonn committed:
fix

jjaassoonn committed:
apply golf by @eric-wieser

jjaassoonn committed:
fix linter error

alreadydone commented on src/algebra/graded_monoid.lean:
```diff
@@ -201,6 +213,34 @@ instance gcomm_monoid.to_comm_monoid [add_comm_monoid ι] [gcomm_monoid A] :
+/-- A graded version of `mul_action`. -/
+class gmul_action [add_monoid ι] [gmonoid A] extends ghas_smul A M :=
+(one_smul (b : graded_monoid M) : (1 : graded_monoid A) • b = b)
+(mul_smul (a a' : graded_monoid A) (b : graded_monoid M) : (a * a') • b = a • a' • b)
```
I wonder why you don't restrict `a, a', b` to some `A i` / `M j` like in `gdistrib_mul_action` and `gmodule` below. Is that not enough?

jjaassoonn replied:
For type checking reasons. For example for `b : M j` `smul 1 b` is of type `M (0 + j)`, so can't be compared with `b` anymore.

eric-wieser replied:
The relevant comparison here is to `gmonoid`.

eric-wieser commented on src/algebra/direct_sum/decomposition.lean:
```diff
@@ -72,6 +72,22 @@ def decompose : M ≃ ⨁ i, ℳ i :=
+      (decomposition.is_internal ℳ')).symm ▸ trivial,
+  exact λ m, add_submonoid.supr_induction ℳ' (mem m) (λ i m h, h_homogeneous ⟨m, h⟩) h_zero h_add,
+end
+
```
This isno longer used by this PR, right?

jjaassoonn replied:
I think so

jjaassoonn committed:
Merge remote-tracking branch 'origin/master' into jjaassoonn/graded_module

eric-wieser commented on src/algebra/graded_monoid.lean:
```diff
@@ -121,11 +121,23 @@ instance ghas_one.to_has_one [has_zero ι] [ghas_one A] : has_one (graded_monoid
+class ghas_smul [has_add ι] :=
+(smul {i j} : A i → M j → M (i + j))
+
+instance ghas_mul.to_ghas_smul [has_add ι] [ghas_mul A] : ghas_smul A A :=
```
```suggestion
/-- A graded version of `has_mul.to_has_smul` -/
instance ghas_mul.to_ghas_smul [has_add ι] [ghas_mul A] : ghas_smul A A :=
```

eric-wieser commented on src/algebra/graded_monoid.lean:
```diff
@@ -201,6 +213,34 @@ instance gcomm_monoid.to_comm_monoid [add_comm_monoid ι] [gcomm_monoid A] :
+(one_smul (b : graded_monoid M) : (1 : graded_monoid A) • b = b)
+(mul_smul (a a' : graded_monoid A) (b : graded_monoid M) : (a * a') • b = a • a' • b)
+
+instance gmonoid.to_gmul_action [add_monoid ι] [gmonoid A] :
```
```suggestion
/-- The graded version of `monoid.to_mul_action`. -/
instance gmonoid.to_gmul_action [add_monoid ι] [gmonoid A] :
```

jjaassoonn committed:
Apply suggestions from code review

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

eric-wieser commented on src/algebra/graded_monoid.lean:
```diff
@@ -448,10 +488,37 @@ instance set_like.ghas_mul {S : Type*} [set_like S R] [has_mul R] [has_add ι] (
 class set_like.graded_monoid {S : Type*} [set_like S R] [monoid R] [add_monoid ι]
   (A : ι → S) extends set_like.has_graded_one A, set_like.has_graded_mul A : Prop
 
+instance set_like.graded_monoid.to_has_graded_smul
```
This should only need `set_like.has_graded_mul`; there should be a stronger version that says `set_like.graded_monoid` implies `gmul_action`.

alreadydone replied:
Is this resolved now?

eric-wieser commented on src/algebra/graded_monoid.lean:
```diff
@@ -448,10 +488,37 @@ instance set_like.ghas_mul {S : Type*} [set_like S R] [has_mul R] [has_add ι] (
+instance set_like.graded_monoid.to_has_graded_smul
+  {S : Type*} [set_like S R] [monoid R] [add_monoid ι] (A : ι → S) [set_like.graded_monoid A] :
+  set_like.has_graded_smul A A :=
+{ smul_mem := λ i j ai bj hi hj, set_like.graded_monoid.mul_mem hi hj, }
```
To elaborate, this is the version with weaker assumptions:
```suggestion
/-- Internally graded version of `has_mul.to_has_smul`. -/
instance set_like.has_graded_mul.to_has_graded_smul
  {S : Type*} [set_like S R] (A : ι → S) [set_like.has_graded_mul A] :
  set_like.has_graded_smul A A :=
{ smul_mem := λ i j ai bj hi hj, set_like.graded_monoid.mul_mem hi hj, }
```

eric-wieser commented on src/algebra/module/graded_module.lean:
```diff
@@ -0,0 +1,189 @@
+  gdistrib_mul_action (λ i, 𝓐 i) (λ i, 𝓜 i) :=
+{ smul := λ i j x y, ⟨(x : A) • (y : M), set_like.has_graded_smul.smul_mem x.2 y.2⟩,
+  one_smul := λ ⟨i, m⟩, sigma.subtype_ext (zero_add _) (one_smul _ _),
+  mul_smul := λ ⟨i, a⟩ ⟨j, a'⟩ ⟨k, b⟩, sigma.subtype_ext (add_assoc _ _ _) (mul_smul _ _ _),
```
These three lines should be used to prove a weaker `gmul_action` instance first.

jjaassoonn replied:
done

jjaassoonn committed:
Update src/algebra/graded_monoid.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

eric-wieser commented on src/algebra/module/graded_module.lean:
```diff
@@ -0,0 +1,189 @@
+variables [add_comm_monoid M] [module A M] [set_like σ M] [add_submonoid_class σ' A]
+  [add_submonoid_class σ M] [set_like.graded_monoid 𝓐] [set_like.has_graded_smul 𝓐 𝓜]
+
+/- [set_like.graded_monoid 𝓐] [set_like.has_graded_smul 𝓐 𝓜] is the internal version of graded
```
```suggestion
/-- `[set_like.graded_monoid 𝓐] [set_like.has_graded_smul 𝓐 𝓜]` is the internal version of graded
```

jjaassoonn committed:
Update src/algebra/module/graded_module.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

jjaassoonn committed:
add `gmul_action`

jjaassoonn committed:
add `gmul_action`

jjaassoonn committed:
fix

jjaassoonn committed:
unused argument fix

kbuzzard commented on src/algebra/module/graded_module.lean:
```diff
@@ -0,0 +1,191 @@
+  map_zero' := add_monoid_hom.ext $ λ a, graded_monoid.gmodule.zero_smul a,
+  map_add' := λ a₁ a₂, add_monoid_hom.ext $ λ b, graded_monoid.gmodule.add_smul _ _ _}
+
+/-- For graded monoid `A` and a graded module `M` over `A`. `gmodule.smul_add_monoid_hom` is the
```
From here to line 114 or so everything is in the `gmodule` namespace, so you could just go into the namespace here I guess.

kbuzzard commented on src/algebra/module/graded_module.lean:
```diff
@@ -0,0 +1,191 @@
+    rw [ring_hom.id_apply, ← direct_sum.sum_support_decompose 𝓜 y, map_sum, finset.smul_sum,
+      map_sum, finset.smul_sum, finset.sum_congr rfl (λ j hj, _)],
+    unfold has_smul.smul,
+    simp only [direct_sum.decompose_add_equiv_apply, direct_sum.decompose_coe,
```
```suggestion
    simp only [(•), direct_sum.decompose_add_equiv_apply, direct_sum.decompose_coe,
      gmodule.smul_add_monoid_hom_apply_of_of],
```

kbuzzard reviewed (approved):
This looks fine to me! I've left some minor comments.

eric-wieser committed:
Add some docstrings, move around definitions

eric-wieser committed:
Split a file

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into jjaassoonn/graded_module

eric-wieser commented on src/algebra/module/graded_module.lean:
```diff
@@ -0,0 +1,218 @@
+  module A (⨁ i, 𝓜 i) :=
+{ smul := λ a b, direct_sum.decompose 𝓐 a • b,
+  ..module.comp_hom _ $
+    (direct_sum.decompose_ring_equiv 𝓐 : A ≃+* ⨁ i, 𝓐 i).to_ring_hom }
```
This one worries me a bit; does `⨁ i, 𝓜 i` not already inherit a different action by `A`? Edit: No, nevermind

alreadydone committed:
fixes

alreadydone committed:
Merge remote-tracking branch 'origin/master' into jyxu/graded_module

alreadydone committed:
address reviews

alreadydone commented on src/algebra/direct_sum/ring.lean:
```diff
@@ -122,6 +122,16 @@ class gcomm_ring [add_comm_monoid ι] [Π i, add_comm_group (A i)] extends
+  smul_zero := λ i j, gsemiring.mul_zero,
+  add_smul := λ i j, gsemiring.add_mul,
+  zero_smul := λ i j, gsemiring.zero_mul,
+  ..graded_monoid.gmonoid.to_gmul_action A }
```
I fixed this error [here](https://github.com/leanprover-community/mathlib/commit/ba994c5b8d4b3b95197bf5dfbba3a391ecfaa36a) by moving this declaration to the graded_module file. If everyone is happy we can merge my branch then send this PR to bors.

alreadydone replied:
Also see the [latest commit](https://github.com/leanprover-community/mathlib/commit/71d4efd74657593064f791206c81f05eb0bf746f) that addressed @kbuzzard's comments.

alreadydone committed:
remove alias of set_like.has_graded_smul.smul_mem

alreadydone committed:
Merge remote-tracking branch 'origin/master' into jyxu/graded_module

jcommelin reviewed (approved):
Thanks :tada:

bors merge
